src/ServicePulse.Host/vue/src/components/failedmessages/FlowDiagram.vue

```diff
@@ -61,292 +58,176 @@ function mapMessage(message) {
+            return {
+              result: [
+                ...result,
+                {
+                  id: `${message.messageId}##${message.receivingEndpoint}`,
+                  type: "message",
+                  data: message,
+                  label: message.nodeName,
+                  position: { x: message.XPos * 300, y: message.level * 200 },
+                },
+              ],
+              currentWidth: currentParentWidth + message.width,
+              previousParent: parentMessage,
+            };
+          },
+          { result: [], currentWidth: 0, previousParent: null }
+        ).result;
+      })
+  );
+}
+
+function constructEdges(mappedMessages) {
+  return mappedMessages
+    .filter((message) => message.parentId)
+    .map((message) => ({
+      id: `${message.parentId}##${message.messageId}`,
+      source: `${message.parentId}##${message.parentEndpoint}`,
+      target: `${message.messageId}##${message.receivingEndpoint}`,
+      markerEnd: MarkerType.ArrowClosed,
+      style: {
+        "stroke-dasharray": message.type === "Event message" && "5, 3",
+      },
+    }));
+}
+
+const elements = ref([]);
+const { onPaneReady, fitView } = useVueFlow();
+
+onMounted(async () => {
+  const messages = await getConversation(props.conversationId);
+  const mappedMessages = messages.map(mapMessage);
+
+  const assignDescendantLevelsAndWidth = (message, level = 0) => {
+    message.level = level;
+    const children = mappedMessages.filter((mm) => mm.parentId === message.messageId && mm.parentEndpoint === message.receivingEndpoint);
+    message.width =
+      children.length === 0
+        ? 1 //leaf node
+        : children.map((child) => (child.width == null ? assignDescendantLevelsAndWidth(child, level + 1) : child)).reduce((sum, { width }) => sum + width, 0);
+    return message;
+  };
+  for (const root of mappedMessages.filter((message) => !message.parentId)) assignDescendantLevelsAndWidth(root);
 
-  for (i = 0; i < messages.length; i += 1) {
-    map[messages[i].messageId] = i; // initialize the map
-    messages[i].children = []; // initialize the children
-  }
+  elements.value = [...constructNodes(mappedMessages), ...constructEdges(mappedMessages)];
+  //TODO: if doing fitView on next Vue onUpdated() then it doesn't appear the elements
+  // are actually drawn yet. See if we can determine a better event to use this on rather than relying on setTimeout
+  setTimeout(() => fitView(), 50);
+});
 
-  for (i = 0; i < messages.length; i += 1) {
-    node = messages[i];
-    if (node.parentId && map[node.parentId]) {
-      // if you have dangling branches check that map[node.parentId] exists
-      messages[map[node.parentId]].children.push(node);
-    } else {
-      roots.push(node);
-    }
-  }
+onPaneReady(({ fitView }) => {
+  fitView();
+});
 
-  return roots;
+function typeIcon(type) {
+  //TODO can this be done without the type magic strings?
```

Comment:
interesting, not sure I can think of a different way

Reply:
I did, will come back to this